pkg/cache/multi_namespace_cache.go

```diff
@@ -37,37 +39,150 @@ type NewCacheFunc func(config *rest.Config, opts Options) (Cache, error)
 // a new global namespaced cache to handle cluster scoped resources.
 const globalCache = "_cluster-scope"
 
-// MultiNamespacedCacheBuilder - Builder function to create a new multi-namespaced cache.
-// This will scope the cache to a list of namespaces. Listing for all namespaces
-// will list for all the namespaces that this knows about. By default this will create
-// a global cache for cluster scoped resource. Note that this is not intended
-// to be used for excluding namespaces, this is better done via a Predicate. Also note that
-// you may face performance issues when using this with a high number of namespaces.
-func MultiNamespacedCacheBuilder(namespaces []string) NewCacheFunc {
+// MultiNamespacedOption is a function that modifies a MultiNamespacedOptions.
+type MultiNamespacedOption func(*MultiNamespacedOptions)
+
+// MultiNamespacedOptions is used to configure the functions used to create caches
+// on a per-namespace basis.
+type MultiNamespacedOptions struct {
+	NewNamespaceCaches        map[string]NewCacheFunc
+	NewClusterScopedCache     NewCacheFunc
+	NewDefaultNamespacedCache NewCacheFunc
+}
+
+// WithLegacyNamespaceCaches configures the MultiNamespacedCacheWithOptionsBuilder
+// with standard caches in each of the namespaces provided as well as for cluster-scoped
+// objects. This option enables use of the MultiNamespacedCacheWithOptionsBuilder
+// to match the behavior of the deprecated MultiNamespacedCacheBuilder.
+func WithLegacyNamespaceCaches(namespaces []string) MultiNamespacedOption {
+	return func(options *MultiNamespacedOptions) {
+		WithNamespaceCaches(namespaces, New)(options)
+		WithClusterScopedCache(New)(options)
+	}
+}
+
+// WithNamespaceCaches configures MultiNamespacedCacheWithOptionsBuilder
+// with namespace-specific caches that are created using the provided NewCacheFunc.
+func WithNamespaceCaches(namespaces []string, f NewCacheFunc) MultiNamespacedOption {
+	return func(options *MultiNamespacedOptions) {
+		for _, ns := range namespaces {
+			WithNamespaceCache(ns, f)(options)
+		}
+	}
+}
+
+// WithNamespaceCache configures MultiNamespacedCacheWithOptionsBuilder
+// with a namespace cache that uses the provided NewCacheFunc.
+func WithNamespaceCache(namespace string, f NewCacheFunc) MultiNamespacedOption {
```

Comment:
`MultiNamespacedCacheWithOptionsBuilder` is 👍 . Just one question, since we are exposing
`WithNamespaceCaches` and `WithNamespaceCache`, what if a user does:
```go
cache.WithNamespaceCache(testNamespaceOne, cache.BuilderWithOptions(....)),
cache.WithNamespaceCaches([]{testNamespaceOne, testNamespaceTwo, testNamespaceThree}, Cache.New),
```
In this case, looks like we will simply override the custom cache func which the user provided for `testNamespaceOne`. Is this something we want to address and validate the inputs, or leave it to the user's responsibility to not give us duplicates?

Reply:
Correct, that would result in `testNamespaceOne: cache.New`

I could document that if the same namespace is provided multiple times, the last applied NewCacheFunc for that namespace wins.

Reply:
Last commit adds a line in the godoc that last setting for a particular namespace wins.